Firely.Fhir.Packages/Core/FhirCiClient.cs

```diff
@@ -0,0 +1,1083 @@
+            }
+
+            return v;
+
+            //return listing is null ? new Versions() : listing.ToVersions();
+        }
+
+        /// <summary>
+        /// Download a package from the source.
+        /// </summary>
+        /// <param name="reference">Package reference of the package to be downloaded.</param>
+        /// <returns>Package content as byte array.</returns>
+        public async Task<byte[]> GetPackage(PackageReference reference)
+        {
+            return await downloadPackage(reference);
+        }
+
+        /// <summary>
+        /// Checks if a package is outdated by comparing its build date with the build date from the QA listings.
+        /// </summary>
+        /// <param name="reference">The package reference.</param>
+        /// <param name="cache">The package cache.</param>
+        /// <returns>True if the package is outdated, false if it is up to date, or null if the package is not found in the QA listings.</returns>
+        public async Task<bool?> IsPackageOutdated(PackageReference reference, IPackageCache cache)
+        {
+            // find our package in the QA listings
+            FhirCiQaRecord? qa = await getQaRecord(reference.Name, reference.Version);
+
+            if (qa == null)
+            {
+                return null;
+            }
+
+            // make sure we have the tag and resolved version references for comparison
+            PackageReference taggedReference = await GetTagBasedReference(reference.Name, reference.Version, qa);
+            PackageReference resolvedReference = await GetResolvedReference(reference.Name, reference.Version, qa);
+
+            // if either fails, we cannot proceed
+            if ((taggedReference == PackageReference.None) ||
+                (resolvedReference == PackageReference.None))
+            {
+                return null;
+            }
+
+            // check for a local copy - CI builds are installed by tag
+            bool isInstalled = await cache.IsInstalled(taggedReference);
+
+            if (!isInstalled)
+            {
+                return true;
+            }
+
+            // get the manifest from the local copy
+            PackageManifest? installedManifest = await cache.ReadManifest(taggedReference);
+
+            if (installedManifest == null)
+            {
+                return true;
+            }
+
+            // compare the build date from QA and the installed manifest
+            return (installedManifest.Date ?? DateTimeOffset.MinValue) < (qa.BuildDateIso ?? qa.BuildDate ?? DateTimeOffset.MaxValue);
+        }
+
+        /// <summary>
+        /// Installs or updates a package from the FHIR CI server.
+        /// </summary>
+        /// <param name="reference">The package reference to install or update.</param>
+        /// <param name="cache">The package cache.</param>
+        public async Task InstallOrUpdate(PackageReference reference, IPackageCache cache)
```

Comment:
This actually looks like a fine function? What bothers you about it?

Reply:
My concern is that it does not use the existing `Install` function from `IPackageCache`.  I am not sure if that should exist there, move into an extension function, or add to the `IPackageCache` interface.